mitmproxy/builtins/script.py

```diff
@@ -0,0 +1,162 @@
+        yield
+    except Exception:
+        _, _, tb = sys.exc_info()
+        scriptdir = os.path.dirname(os.path.abspath(path))
+        for i, s in enumerate(reversed(traceback.extract_tb(tb))):
+            if not os.path.abspath(s[0]).startswith(scriptdir):
+                break
+            else:
+                tb = tb.tb_next
+        ctx.log.warn("".join(traceback.format_tb(tb)))
+    finally:
+        sys.argv = oldargs
+        sys.path.pop()
+
+
+def load_script(path, args):
+    ns = {'__file__': os.path.abspath(path)}
+    with scriptenv(path, args):
+        with open(path, "rb") as f:
+            code = compile(f.read(), path, 'exec')
+            exec(code, ns, ns)
+    return ns
+
+
+class ReloadHandler(watchdog.events.FileSystemEventHandler):
+    def __init__(self, callback):
+        self.callback = callback
+
+    def on_modified(self, event):
+        self.callback()
+
+    def on_created(self, event):
+        self.callback()
+
+
+class Script:
+    """
+        An addon that manages a single script.
+    """
+    def __init__(self, command):
+        self.name = command
+
+        self.command = command
+        self.path, self.args = parse_command(command)
+        self.ns = None
+        self.observer = None
+
+        self.last_options = None
+        self.should_reload = threading.Event()
+
+        for i in controller.Events - set(["tick"]):
+            def mkprox():
+                evt = i
+
+                def prox(*args, **kwargs):
+                    self.run(evt, *args, **kwargs)
+                return prox
+            setattr(self, i, mkprox())
+
+    def run(self, name, *args, **kwargs):
+        # It's possible for ns to be un-initialised if we failed during
+        # configure
+        if self.ns is not None:
+            func = self.ns.get(name)
+            if func:
+                with scriptenv(self.path, self.args):
+                    func(*args, **kwargs)
+
+    def reload(self):
+        self.should_reload.set()
+
+    def tick(self):
+        if self.should_reload.is_set():
+            self.should_reload.clear()
+            self.ns = None
+            ctx.log.info("Reloading script: %s" % self.name)
+            self.configure(self.last_options)
+        else:
+            self.run("tick")
+
+    def configure(self, options):
+        self.last_options = options
+        if not self.observer:
+            self.observer = Observer()
+            # Bind the handler to the real underlying master object
+            self.observer.schedule(
+                ReloadHandler(self.reload),
+                os.path.dirname(self.path) or "."
+            )
+            self.observer.start()
+        if not self.ns:
+            self.ns = load_script(self.path, self.args)
+        self.run("configure", options)
+
+
+class ScriptLoader():
+    """
+        An addon that manages loading scripts from options.
+    """
+    def configure(self, options):
+        for s in options.scripts or []:
```

Comment:
Per https://github.com/mitmproxy/mitmproxy/blob/e051928f26afbc58c01d7854b62f8bfb0d4ed5ee/mitmproxy/flow/options.py#L24, `.scripts` is never None and always an empty sequence.